R/prepare_stan_data_inputs.R

```diff
@@ -22,49 +23,242 @@
 
   ## Covariate additions
   if (!is.null(analysis_obj@covariates)) {
     data_in[["K"]] <- NROW(analysis_obj@covariates@covariates)
-    data_in[["X"]] <- trimmed_data_matrix[, analysis_obj@covariates@covariates, drop = FALSE]
+    data_in[["X"]] <- data_matrix[, analysis_obj@covariates@covariates, drop = FALSE]
     beta_constraints <- get_covariate_constraints(analysis_obj@covariates)
     data_in[["L_beta"]] <- beta_constraints[, "lower"]
     data_in[["U_beta"]] <- beta_constraints[, "upper"]
   }
 
   ## Weights
   if (analysis_obj@outcome@weight_var != "") {
-    data_in[["weight"]] <- trimmed_data_matrix[, analysis_obj@outcome@weight_var]
+    data_in[["weight"]] <- data_matrix[, analysis_obj@outcome@weight_var]
   }
 
   return(data_in)
 }
+
+# class union ----
+setClassUnion("OutcomeSurvExponentialWeibull", c("OutcomeSurvExponential", "OutcomeSurvWeibullPH"))
+
+# Survival ----
+## Exponential / Weibull ----
+### No/full borrowing ----
+setMethod(
+  f = "prepare_stan_data_inputs",
+  signature = c("OutcomeSurvExponentialWeibull", "BorrowingNoneFull", "ANY"),
+  definition = function(outcome, borrowing, analysis_obj) {
+    data_matrix <- trim_data_matrix(analysis_obj)
+    data_in <- list(
+      N = nrow(data_matrix),
+      trt = data_matrix[, analysis_obj@treatment@trt_flag_col],
+      time = data_matrix[, outcome@time_var],
+      cens = data_matrix[, outcome@cens_var]
+    )
+
+    # Add covariates and weights
+    data_in <- add_covariates_and_weights(data_in, analysis_obj, data_matrix)
+
+    return(data_in)
+  }
+)
+
+### Hierarchical commensurate prior borrowing ----
+setMethod(
+  f = "prepare_stan_data_inputs",
+  signature = c("OutcomeSurvExponentialWeibull", "BorrowingHierarchicalCommensurate", "ANY"),
+  definition = function(outcome, borrowing, analysis_obj) {
+    data_matrix <- analysis_obj@data_matrix
+    data_in <- list(
+      N = nrow(data_matrix),
+      trt = data_matrix[, analysis_obj@treatment@trt_flag_col],
+      time = data_matrix[, outcome@time_var],
+      cens = data_matrix[, outcome@cens_var],
+      Z = cbind(
+        1 - data_matrix[, borrowing@ext_flag_col],
+        data_matrix[, borrowing@ext_flag_col]
+      )
+    )
+
+    # Add covariates and weights
+    data_in <- add_covariates_and_weights(data_in, analysis_obj, data_matrix)
+
+    return(data_in)
+  }
+)
+
+# Survival ----
+## PEM ----
+### No/full borrowing ----
+setMethod(
+  f = "prepare_stan_data_inputs",
+  signature = c("OutcomeSurvPEM", "BorrowingNoneFull", "ANY"),
+  definition = function(outcome, borrowing, analysis_obj) {
+
+    analysis_obj@data_matrix <- trim_data_matrix(analysis_obj)
+    analysis_obj <- cast_mat_to_long_pem(analysis_obj)
+    data_matrix <- analysis_obj@data_matrix
+
+    n_periods <- analysis_obj@outcome@n_periods
+    Z <- matrix(0, nrow = nrow(data_matrix), ncol = n_periods)
+    for (i in 1:nrow(data_matrix)) {
+      period <- data_matrix[i, "__period__"]
+      Z[i, period] <- 1
+    }
```

Comment:
how about a cryptic one-liner?
```suggestion
    Z[matrix[c(seq_len(nrow(data_matrix)), data_matrix[, "__period__"]), ncol = 2]] <- 1    
```

Could also define it as
```r
indicator_location <- matrix[c(seq_len(nrow(data_matrix)), data_matrix[, "__period__"]), ncol = 2]
Z[indication_location] <- 1
```